rdflib/graph.py

```diff
@@ -43,9 +43,8 @@
 
 logger = logging.getLogger(__name__)
 
-# Type aliases to make unpacking what's going on a little more human friendly
-ContextNode = Union[BNode, URIRef]
-DatasetQuad = Tuple[Node, URIRef, Node, Optional[ContextNode]]
+# Type aliase to make unpacking what's going on a little more human friendly
+DatasetQuad = Tuple[Node, URIRef, Node, Optional[IdentifiedNode]]
```

Comment:
There is an issue with the extant typing of `ContextNode` which this change understandably doesn't address. The Graph identifier is (correctly) [type-hinted as `Optional[Union[Node, str]]`](https://github.com/RDFLib/rdflib/blob/b2cae90a34edb1c2a04156ed3e47752e7e161b5f/rdflib/graph.py#L332) and `Optional[IdentifiedNode]` doesn't capture the `str` component.

Reply:
I think this last point is correct: any of those things can be used! I'll add it to the branch.

Reply:
I certainly missed this use case `ContextNode` was supporting, apparently incompletely.  I suppose the question now is, is `ContextNode` intended to be a symbol meant for export and re-consumption, or is it internal to the module?  If internal, would it be better suited as being named `_ContextNode`?

Reply:
Looks like it should be `ContextNode[Any]` :smile:  ...
```py
def test_graph_identifier():
    g1 = Graph(identifier="⚠️")
    logger.debug(f"{g1.identifier}")

    g2 = Graph(identifier=Literal("foo"))
    logger.debug(f"{g2.identifier}")

    g3 = Graph(identifier=BNode())
    logger.debug(f"{g3.identifier}")

    g4 = Graph(identifier=URIRef("urn:example:context-1"))
    logger.debug(f"{g4.identifier}")

    g5 = Graph(identifier=FOAF.name)
    logger.debug(f"{g5.identifier}")

    g6 = Graph(identifier=Namespace("http://rdflib.net"))
    logger.debug(f"{g6.identifier}")

    # At least this is derefenced, phew!
    g7 = Graph(identifier=g1)
    logger.debug(f"{g7.identifier}")
    # lol
    g8 = Graph(identifier=(tarek, likes, pizza))
    logger.debug(f"{g8.identifier}")
```
```
⚠️
foo
N487b9bd4146e44e8b039eadca3a36ec8
urn:example:context-1
http://xmlns.com/foaf/0.1/name
http://rdflib.net
N3b7d314ec941418d9d0f0f6301846796
(rdflib.term.URIRef('urn:example:tarek'), 
 rdflib.term.URIRef('urn:example:likes'), 
 rdflib.term.URIRef('urn:example:pizza'))
PASSED                                                                   [100%]
```
Anything that returns a  `str` from `repr` is valid, so most of the successes are unsurprising.

Reply:
> Meh, I forgot that Graph also accepts a Literal as an identifier, also not captured by `IdentifiedNode`.

It is important to note, this type alias is only used in one place, which is the return value for `Dataset.__iter__` so it should be correct for that, and for this reason it probably should not be a type alias, or at the very least not one outside of the class. At the very least a type alias on module level should be used in more than one class/function in that module before it is clear that it makes sense for it to be an alias, as there may be cases where something looks like a ContextNode and needs to have `Literal`, `str`, `Node` and `None` - and other places where `IdentifiedNode` really is the most appropriate.


Also Literal is not in the value space of graph identifiers/names according to [RDF 1.1](https://www.w3.org/TR/rdf11-concepts/#section-dataset):

> An RDF dataset is a collection of [RDF graphs](https://www.w3.org/TR/rdf11-concepts/#dfn-rdf-graph), and comprises:
> 
> - Exactly one default graph, being an [RDF graph](https://www.w3.org/TR/rdf11-concepts/#dfn-rdf-graph). The default graph does not have a name and may be empty.
> - Zero or more named graphs. Each named graph is a pair consisting of an [IRI](https://www.w3.org/TR/rdf11-concepts/#dfn-iri) or a blank node (the graph name), and an [RDF graph](https://www.w3.org/TR/rdf11-concepts/#dfn-rdf-graph). Graph names are unique within an RDF dataset.

So if Graph accepts it as an identifier, it probably should not, and the right thing is to narrow the type there to match the correct return type for `Dataset.__iter__` - if `Dataset.__iter__` does return something that is contrary to the spec I think we should at least understand where things go awry also.

Reply:
> There is an issue with the extant typing of `ContextNode` which this change understandably doesn't address. The Graph identifier is (correctly) [type-hinted as `Optional[Union[Node, str]]`](https://github.com/RDFLib/rdflib/blob/b2cae90a34edb1c2a04156ed3e47752e7e161b5f/rdflib/graph.py#L332) and `Optional[IdentifiedNode]` doesn't capture the `str` component

as noted elsewhere, Graph will convert any identifier that is not a Node to a URIRef:

https://github.com/RDFLib/rdflib/blob/4eba1a098698f301e2996fba23a72927fb5d07a2/rdflib/graph.py#L342-L345

Reply:
> Looks like it should be `ContextNode[Any]` smile ...

This may be the case for Graph constructor, but all of these values are converted to URIRef if they are not URIRefs:

```console
$ pipx run --spec git+https://github.com/RDFLib/rdflib.git@master#egg=rdflib python
⚠️  python is already on your PATH and installed at /usr/bin/python. Downloading and running anyway.
Python 3.10.2 (main, Jan 17 2022, 00:00:00) [GCC 11.2.1 20211203 (Red Hat 11.2.1-7)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import rdflib
>>> rdflib.Graph(identifier=(1,2,3)).identifier
rdflib.term.URIRef('(1, 2, 3)')
```

And even though it does work in some sense (I'm actually not sure how valid it will be as a IRI, but maybe it is entirely valid), I still think there is an open question as to whether we want to advertise that anything should be passible into it, and certainly there are some things that do fail and it seems tuple works for reasons that were not quite intended:

```console
>>> rdflib.Graph(identifier=3).identifier
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/iwana/.local/pipx/.cache/1500787fc0bbcf9/lib64/python3.10/site-packages/rdflib/graph.py", line 345, in __init__
    self.__identifier = URIRef(self.__identifier)  # type: ignore[unreachable]
  File "/home/iwana/.local/pipx/.cache/1500787fc0bbcf9/lib64/python3.10/site-packages/rdflib/term.py", line 273, in __new__
    if not _is_valid_uri(value):
  File "/home/iwana/.local/pipx/.cache/1500787fc0bbcf9/lib64/python3.10/site-packages/rdflib/term.py", line 96, in _is_valid_uri
    if c in uri:
TypeError: argument of type 'int' is not iterable

```

Reply:
And note that even though things don't throw an exception if someone does not adhere to type annotations, I think the point of type annotations should be to tell people what the correct type is to pass in, so that they can spot bugs in their code, if for some reason things still mostly work in one place when they pass in wrong values, it may still break down later on, and this is basically also declaring our contract with users of the library, if we say we will accept tuples as input to Graph identifier, just because they work, then we have to bear the consequences of this everywhere and maintain this, while if we say that we only accept strings, if someone then passes something else, and it works, well fine, but if something breaks further down the line at least we can tell them that they should be passing in a string instead of a tuple.

Reply:
Don't even need to keep `str`, you are correct, all the sane values resolve to URIRef. I worked up a more thorough check:
```py
import pytest
from rdflib import Namespace, FOAF
from rdflib.graph import Dataset, Graph
from rdflib.term import (
    BNode,
    Literal,
    RDFLibGenid,
    URIRef,
    Node,
    IdentifiedNode,
    Variable,
)

tarek = URIRef("urn:example:tarek")
likes = URIRef("urn:example:likes")
pizza = URIRef("urn:example:pizza")

def test_graph_identifier_as_string():
    g = Graph(identifier="xxx")
    assert type(g.identifier) is URIRef  # rdflib.term.URIRef('xxx')
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_literal():
    g = Graph(identifier=Literal("xxx"))
    assert type(g.identifier) is Literal  # rdflib.term.Literal('xxx')
    assert issubclass(type(g.identifier), Node)
    assert not issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_bnode():
    g = Graph(identifier=BNode())
    assert type(g.identifier) is BNode  # rdflib.term.BNode('Ndc8ac01941254b299a66c31a5b49cdd3')
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_namespace():
    g = Graph(identifier=(tarek, likes, pizza))
    assert type(g.identifier) is URIRef  # rdflib.term.URIRef("(rdflib.term.URIRef('urn:example:tarek'), 
                                         # rdflib.term.URIRef('urn:example:likes'),
                                         # rdflib.term.URIRef('urn:example:pizza'))")
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_graph():
    g = Graph()
    g = Graph(identifier=g)
    assert type(g.identifier) is BNode  # rdflib.term.BNode('N666b78c69a3d4544a079ab0919a84dda')
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_genid():
    g = Graph()
    g = Graph(identifier=RDFLibGenid("xxx"))
    assert type(g.identifier) is RDFLibGenid  # RDFLibGenid('xxx')
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_identifiednode():
    g = Graph()
    g = Graph(identifier=IdentifiedNode("xxx"))
    assert type(g.identifier) is IdentifiedNode   # 'xxx'
    assert issubclass(type(g.identifier), Node)
    assert issubclass(type(g.identifier), IdentifiedNode)

def test_graph_identifier_as_variable():
    g = Graph()
    g = Graph(identifier=Variable("x"))
    assert type(g.identifier) is Variable  # rdflib.term.Variable('x')
    assert issubclass(type(g.identifier), Node)
    assert not issubclass(type(g.identifier), IdentifiedNode)
```